### PR TITLE
rgw: add support for noncurrentversion expiration in s3 lifecycle.

### PR DESCRIPTION
**Pull request by zhangsw**
Add the support for the non-current object version's expiration. This works if the bucket's versioning status is enabled or suspended.
S3's doc about the expiration: http://docs.aws.amazon.com/AmazonS3/latest/dev/intro-lifecycle-rules.html

Fixes: http://tracker.ceph.com/issues/18916

Signed-off-by: Zhang Shaowen <zhangshaowen@cmss.chinamobile.com>

### CODE REVIEW

**Comment by idealguo** [comment]
@yehudasa @cbodley @mattbenjamin, please help to review, thanks

File: src/rgw/rgw_lc.cc [hunk 1]
```diff
@@ -377,9 +339,95 @@ int RGWLC::bucket_lc_process(string& shard_id)
...
+          ldout(cct, 0) << "ERROR: store->list_objects():" <<dendl;
+          return ret;
+        }
+
+        utime_t now = ceph_clock_now();
+        ceph::real_time mtime;
+        bool remove_indeed = true;
+        int expiration;
+        for (auto obj_iter = objs.begin(); obj_iter != objs.end(); ++obj_iter) {
+          if ((*obj_iter).is_current()) {
```
Comment (dang) [inline src/rgw/rgw_lc.cc]: Rather than (*obj_iter).foo, can you use obj_iter->foo

**Comment by cbodley** [comment]
[teuthology run](http://pulpito.ceph.com/cbodley-2017-02-15_16:05:58-rgw-wip-cbodley-testing---basic-mira/) was clean (only failures were selinux related)

**Comment by zhangsw** [comment]
@dang I've changed it to '->'.

File: src/rgw/rgw_lc.cc (hunk 1 above)
Comment (zhangsw) [inline src/rgw/rgw_lc.cc]: @dang Done.

**Comment by zhangsw** [comment]
@dang I've changed to '->'.

**Comment by yehudasa** [comment]
@zhangsw @dang we should make sure to add a test s3-tests to cover this feature

**Pull request merged**